go-controller/pkg/ovn/namespace.go

```diff
@@ -218,7 +236,60 @@ func (oc *Controller) updateNamespace(old, newer *kapi.Namespace) {
 	}
 	defer nsInfo.Unlock()
 
-	annotation := newer.Annotations[hotypes.HybridOverlayExternalGw]
+	var annotation, oldAnnotation string
+	annotation = newer.Annotations[routingExternalGWsAnnotation]
+	oldAnnotation = old.Annotations[routingExternalGWsAnnotation]
+	if annotation != oldAnnotation {
+		var stderr string
+		var err error
+		for k, v := range nsInfo.podExternalRoutes {
+			_, stderr, err = util.RunOVNNbctl("--if-exists", "--policy=src-ip",
+				"lr-route-del", v[0].gr, k+"/32")
+			if err != nil {
+				klog.Errorf("Unable to delete src-ip route to GR router, stderr:%q, err:%v", stderr, err)
+			} else {
+				delete(nsInfo.podExternalRoutes, k)
+			}
+		}
+		if annotation != "" {
+			annotationOk := true
+			nsInfo.routingExternalGWs = nil
+			for _, v := range strings.Split(annotation, ",") {
+				parsedAnnotation := net.ParseIP(v)
+				if parsedAnnotation == nil {
+					klog.Errorf("Could not parse routing external gw annotation")
+					annotationOk = false
+					nsInfo.routingExternalGWs = nil
+					break
+				} else {
+					nsInfo.routingExternalGWs = append(nsInfo.routingExternalGWs, parsedAnnotation)
+				}
+			}
+			if annotationOk {
+				existingPods, err := oc.watchFactory.GetPods(old.Name)
+				if err != nil {
+					klog.Errorf("Failed to get all the pods (%v)", err)
+				} else {
+					for _, pod := range existingPods {
+						podIP := pod.Status.PodIP
+						gr := "GR_" + pod.Spec.NodeName
+						for _, gw := range nsInfo.routingExternalGWs {
+							_, stderr, err = util.RunOVNNbctl("--may-exist", "--policy=src-ip", "--ecmp",
+								"lr-route-add", gr, podIP+"/32", gw.String())
+							if err != nil {
+								klog.Errorf("Unable to add src-ip route to GR router, stderr:%q, err:%v", stderr, err)
+							} else {
+								nsInfo.podExternalRoutes[podIP] = append(nsInfo.podExternalRoutes[podIP], externalRoute{gr: gr, gw: gw.String()})
```

Comment:
instead of podExternalRoutes holding a slice of externalRoutes, could it instead just hold a map of gw.String and gr?
a gw ip can only be specified once for a pod right? so map would be more simple I think

Reply:
done